### PR TITLE
[GPU Process] [FormControls] Enable drawing ControlPart in GPU Process

### PR DESCRIPTION
**Pull request by shallawa**
#### 64bc8ac412efea0e6523fd0f285d2a70c734ffde
<pre>
[GPU Process] [FormControls] Enable drawing ControlPart in GPU Process
<a href="https://bugs.webkit.org/show_bug.cgi?id=249474">https://bugs.webkit.org/show_bug.cgi?id=249474</a>
rdar://103446772

Reviewed by Simon Fraser.

Make DisplayList::Recorder override GraphicsContext::drawControlPart(). Create
a new DisplayList item named DrawControlPart and new RemoteDisplayListRecorder
message named DrawControlPart. Handle the encoding and decoding of the super
classes of ControlPart.

Because GPUP can render ControlParts simultaneously in multiple threads, a
ControlPartFactory is created for every RemoteDisplayListRecorder which handles
drawing the DisplayList item DrawControlPart.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawControlPart):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/controls/ControlPartType.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.cpp: Added.
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/PlatformControl.h:
(WebCore::PlatformControl::PlatformControl):
* Source/WebCore/platform/graphics/displaylists/DisplayList.cpp:
(WebCore::DisplayList::DisplayList::append):
* Source/WebCore/platform/graphics/displaylists/DisplayListItemBuffer.cpp:
(WebCore::DisplayList::ItemHandle::apply):
(WebCore::DisplayList::ItemHandle::destroy):
(WebCore::DisplayList::ItemHandle::safeCopy const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItemType.cpp:
(WebCore::DisplayList::sizeOfItemInBytes):
(WebCore::DisplayList::isDrawingItem):
(WebCore::DisplayList::isInlineItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItemType.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawControlPart::DrawControlPart):
(WebCore::DisplayList::DrawControlPart::apply):
(WebCore::DisplayList::operator&lt;&lt;):
(WebCore::DisplayList::dumpItem):
(WebCore::DisplayList::dumpItemHandle):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawControlPart::type const):
(WebCore::DisplayList::DrawControlPart::rect const):
(WebCore::DisplayList::DrawControlPart::deviceScaleFactor const):
(WebCore::DisplayList::DrawControlPart::style const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::drawControlPart):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordDrawControlPart):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::ControlMac):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h:
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
(WebCore::MeterMac::MeterMac):
* Source/WebCore/rendering/GlyphDisplayListCache.cpp:
(WebCore::canShareDisplayListWithItem):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawControlPart):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode):
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawControlPart):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: <a href="https://commits.webkit.org/258070@main">https://commits.webkit.org/258070@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/42d79c982b767ffe3dc74ce6468e275021072aa9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100825 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9973 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110127 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170401 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104814 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10906 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/757 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93233 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107972 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8242 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34859 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/104349 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90151 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22894 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77834 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3665 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24415 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3684 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9795 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43915 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5535 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5462 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebKit/Platform/IPC/ControlPartReference.h [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ */
+
+#pragma once
+
+#include <WebCore/ControlPart.h>
+#include <WebCore/MeterPart.h>
+
+namespace IPC {
+
+class ControlPartReference {
```
Comment (pxlcoder) [inline Source/WebKit/Platform/IPC/ControlPartReference.h]: Can you explain why a separate class is needed to encode `ControlPart`? I think this is similar to `SystemImage` and it's various `types` – which doesn't have this kind of wrapper.

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp [hunk 2]
```diff
@@ -421,6 +421,19 @@ void ClearRect::apply(GraphicsContext& context) const
     context.clearRect(m_rect);
 }
 
+DrawControlPart::DrawControlPart(const ControlPart& part, const FloatRect& rect, float deviceScaleFactor, const ControlStyle& style)
```
Comment (pxlcoder) [inline Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp]: Should we drop the `const` here, and avoid the `const_cast`? None of the other display list items are written this way.

File: Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp [hunk 3]
```diff
@@ -391,6 +392,11 @@ void RemoteDisplayListRecorderProxy::recordClearRect(const FloatRect& rect)
     send(Messages::RemoteDisplayListRecorder::ClearRect(rect));
 }
 
+void RemoteDisplayListRecorderProxy::recordDrawControlPart(const ControlPart& part, const FloatRect& rect, float deviceScaleFactor, const ControlStyle& style)
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp]: Same comment about `const` and `const_cast` here.

File: Source/WebKit/Platform/IPC/ControlPartReference.h (hunk 1 above)
Comment (shallawa) [reply Source/WebKit/Platform/IPC/ControlPartReference.h]: There are different ways for IPC encoding and decoding the RefCounted objects. And there is no preferred way for every case. At the time I was adding the FilterReference for encoding and decoding the Filter and FilterEffect, I think adding Ref<> in the messages.in files was not allowed. So I followed the pattern of `IPC::FormDataReference` which still exists. 

But you are right! Having `Ref<ControlPart>` in the messages.in file and in the receiver class is much better than `IPC::ControlPartReference`.

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp (hunk 2 above)
Comment (shallawa) [reply Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp]: I added the const qualifier per Simon request in https://github.com/WebKit/WebKit/pull/7566#discussion_r1050011917. It seems it is (and will be) causing trouble. So I removed it.

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp [hunk 4]
```diff
@@ -860,6 +874,12 @@ void dumpItem(TextStream& ts, const ClearRect& item, OptionSet<AsTextFlag>)
     ts.dumpProperty("rect", item.rect());
 }
 
+void dumpItem(TextStream& ts, const DrawControlPart& item, OptionSet<AsTextFlag>)
+{
+    ts.dumpProperty("type", item.type());
+    ts.dumpProperty("rect", item.rect());
```
Comment (smfr) [inline Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:880]: You will probably help your future self by dumping the `ControlStyle` and m_deviceScaleFactor

File: Source/WebCore/platform/graphics/GraphicsContext.cpp [hunk 5]
```diff
@@ -374,7 +374,7 @@ void GraphicsContext::drawPattern(ImageBuffer& image, const FloatRect& destRect,
     image.drawPattern(*this, destRect, tileRect, patternTransform, phase, spacing, options);
 }
 
-void GraphicsContext::drawControlPart(const ControlPart& part, const FloatRect& rect, float deviceScaleFactor, const ControlStyle& style)
+void GraphicsContext::drawControlPart(ControlPart& part, const FloatRect& rect, float deviceScaleFactor, const ControlStyle& style)
```
Comment (smfr) [inline Source/WebCore/platform/graphics/GraphicsContext.cpp:377]: I'd like to find out if we can keep the ControlPart& `const`. Drawing one should not mutate it.

**Review by smfr — APPROVED** [review]

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp (hunk 4 above)
Comment (shallawa) [reply Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:880]: Fixed.

**Pull request merged**